internal/jobloop/tx_guarded.go

```diff
@@ -0,0 +1,106 @@
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+******************************************************************************/
+
+package jobloop
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sapcc/go-bits/sqlext"
+)
+
+// TxGuardedJob is a specialization of ProducerConsumerJob, where each task
+// corresponds to one row of a database table that needs to be worked on.
+// Rows must be selected in a ConcurrencySafe way (most commonly through the
+// "SELECT ... FOR UPDATE SKIP LOCKED" mechanism). The job implementation
+// ensures that the entirety of each task runs within a single SQL transaction,
+// and manages the lifecycle of that transaction.
+//
+// This type can be used in the same way as ProducerConsumerJob, except that it
+// offers a different set of callbacks. The type arguments are:
+//   - P (Payload), the payload for one individual task (e.g. the ORM object
+//     corresponding to the selected row)
+//   - Tx (Transaction), the type for a DB transaction (the job implementation
+//     will call Rollback on this in case of errors)
+type TxGuardedJob[Tx sqlext.Rollbacker, P any] struct {
+	Metadata JobMetadata
+
+	// A function that begins a new DB transaction. Usually set to `db.Begin`.
+	BeginTx func() (Tx, error)
+	// A function that will be polled periodically (once per transaction) to
+	// discover the next row to work on. If there are currently no rows waiting
+	// to be processed, this function shall return `sql.ErrNoRows` to instruct
+	// the job to slow down its polling.
+	//
+	// The provided label set will have been prefilled with the labels from
+	// Metadata.CounterLabels and all label values set to "early-db-access". The
+	// implementation is expected to substitute the actual label values as soon
+	// as they become known.
+	DiscoverRow func(Tx, prometheus.Labels) (P, error)
+	// A function that will be called once for each discovered row to process it.
+	//
+	// The provided label set will have been prefilled with the labels from
+	// Metadata.CounterLabels and all label values set to "early-db-access". The
+	// implementation is expected to substitute the actual label values as soon
+	// as they become known.
+	ProcessRow func(Tx, P, prometheus.Labels) error
+}
+
+// Setup builds the Job interface for this job and registers the counter
+// metric. At runtime, `nil` can be given to use the default registry. In
+// tests, a test-local prometheus.Registry instance should be used instead.
```

Comment:
Is that a thing that you quote nil but things like `prometheus.Registry` not?

Reply:
I know that I'm not entirely consistent about this. Traditionally, Go documentation is just plaintext, though they're adding some Markdown support soon (in Go 1.21, if I remember correctly). I tend to put backticks around things that are code snippets if it feels like it would be more readable like this. For this `nil` specifically, it makes it stand out as a code keyword vs. natural language word. For `prometheus.Registry`, there would not be any confusion since it is very obviously a type name.